src/ucp/core/ucp_worker.c

```diff
@@ -2350,3 +2356,90 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
 
     UCP_WORKER_THREAD_CS_EXIT_CONDITIONAL(worker);
 }
+
+static int ucp_worker_keepalive_is_enabled(ucp_worker_h worker)
+{
+    return (worker->context->config.ext.keepalive_timeout != 0.0) &&
+           (worker->context->config.ext.keepalive_num_eps != 0);
+}
+
+static unsigned ucp_worker_keepalive_cb(void *arg)
+{
+    ucp_worker_h worker = (ucp_worker_h)arg;
+    double now          = ucs_get_accurate_time();
+    unsigned cnt;
+    ucp_ep_ext_gen_t *ep_ext;
+    ucs_list_link_t *iter_begin;
+
+    if (ucs_likely((now - worker->keepalive.last_shot) <
+                   worker->context->config.ext.keepalive_timeout)) {
+        return 0;
+    }
+
+    worker->keepalive.last_shot = now;
+    cnt                         = 0;
+    iter_begin                  = worker->keepalive.iter;
+    /* use own loop for elements because standard for_each skips
+     * head element */
+    do {
+        ep_ext = ucs_list_next(worker->keepalive.iter,
+                               ucp_ep_ext_gen_t, ep_list);
+        worker->keepalive.iter = &ep_ext->ep_list;
+        if (worker->keepalive.iter == &worker->all_eps) {
+            continue;
+        }
+
+        cnt += !!ucp_ep_do_keepalive(ucp_ep_from_ext_gen(ep_ext));
+        if (cnt >= worker->context->config.ext.keepalive_num_eps) {
+            break;
+        }
+    } while (iter_begin != &ep_ext->ep_list);
+
+    if (cnt == 0) {
```

Comment:
what if all eps in the list did not have resources to send `ep_check`?
We should not unsubscribe in this case

Reply:
right, but on ep reconfiguration it will be enabled again

Reply:
ok